bip-0324.mediawiki

```diff
@@ -0,0 +1,467 @@
+
+After a successful handshake, messages MUST use the "v2 messages structure". Non-encrypted v1 messages from the initiating peer MUST lead to an immediate connection termination.
+
+After a successful handshake, both peers MUST wipe the ephemeral-session-key from memory and/or persistence storage.
+
+A peer not supporting this proposal will not perform the described handshake and thus send a v1 version message.
+Peers supporting this BIP MAY optionally allow unencrypted v1 communication by detecting a v1 version message by the initial 11-byte sequence of <code>4byte net magic || "version"</code>.
+
+=== Symmetric Encryption Cipher Keys ===
+
+Once the ECDH secret (<code>ECDH_KEY</code>) is calculated on each side, the symmetric encryption cipher keys MUST be derived with HKDF <ref>[https://tools.ietf.org/html/rfc5869 HKDF (RFC 5869)]</ref> after the following specification:
+
+1. HKDF extraction
+<code>PRK = HKDF_EXTRACT(hash=SHA256, salt="BitcoinSharedSecret||INITIATOR_32BYTES_PUBKEY||RESPONDER_32BYTES_PUBKEY||NETWORK_MAGIC", ikm=ECDH_KEY)</code>.
+
+2. Derive Key_1_A (K_1 communication direction A)
+<code>K1A = HKDF_EXPAND(prk=PRK, hash=SHA256, info="BitcoinK_1_A", L=32)</code>
+
+2. Derive Key_2_A (K_2 communication direction A)
+<code>K1B = HKDF_EXPAND(prk=PRK, hash=SHA256, info="BitcoinK_2_A", L=32)</code>
+
+3. Derive Key_1_B (K_1 communication direction B)
+<code>K2 = HKDF_EXPAND(prk=PRK, hash=SHA256, info="BitcoinK_1_B", L=32)</code>
+
+3. Derive Key_2_B (K_2 communication direction B)
+<code>K2 = HKDF_EXPAND(prk=PRK, hash=SHA256, info="BitcoinK_2_B", L=32)</code>
+
+=== Session ID ===
+
+Both parties MUST also calculate the 256bit session-id using <code>SID = HKDF_EXPAND(prk=PRK, hash=SHA256, info="BitcoinSessionID", L=32)</code>. The session-id can be used for authenticating the encryption-session (identity check).
+
+The session-id MUST be presented to the user on request.
+
+=== ChaCha20-Poly1305@Bitcoin Cipher Suite ===
+
+==== Background ====
+
+ChaCha20-Poly1305@Bitcoin AEAD is almost identical to the openSSH version <ref>[https://github.com/jhcloos/openssh-chacha-poly1305/blob/master/PROTOCOL.chacha20poly1305]</ref>. The only difference to the @openSSH version is that the @Bitcoin version has a 3 byte package length (instead 4) and reuses the remaining bytes of the ChaCha20 round (21 length field encryption per ChaCha20 round; 21 x 3 < 64).
+
+ChaCha20 is a stream cipher designed by Daniel Bernstein and described in <ref>[http://cr.yp.to/chacha/chacha-20080128.pdf ChaCha20]</ref>. It operates by permuting 128 fixed bits, 128 or 256 bits of key, a 64 bit nonce and a 64 bit counter into 64 bytes of output. This output is used as a keystream, with any unused bytes simply discarded.
```

Comment:
"any unused bytes simply discarded" -> "unused bytes discarded".

"128 or 256 bits". Surely this must be fixed for the purposes of this spec? If one of these configurations is unused in this BIP, maybe it would be nice to explain which one it is. If they can somehow both be used, it would be nice with a pointer to the section where this is selected.